docs/proposals/20200304-autoscaler-scale-ux.md

```diff
@@ -0,0 +1,117 @@
+---
+title: Cluster Autoscaler Scale from Zero UX
+authors:
+  - "@michaelgugino"
+reviewers:
+  - "@detiber"
+creation-date: 2020-03-04
+creation-date: 2020-03-04
+status:
+---
+
+# Cluster Autoscaler Scale from Zero UX
+
+## Summary
+Cluster Autoscaler supports scale from zero for cloud-native providers. This is done via compiling CPU, GPU, and RAM info into the cluster autoscaler at build time for each supported cloud.
+
+We are attempting to add cluster-api support to the cluster-autoscaler to scale machinesets, machinedeployments, and/or other groupings of machines as appropriate. We are trying to keep an abstraction which is isolated from particular cloud implementations. We need a way to inform the cluster-autoscaler of the value of machine resources so it can make proper scheduling decisions when there are not currently any machines in a particular scale set (machineset/machinedeployment, etc).
+
+We will need input from the autoscaler team, as well as input from api reviewers to ensure we obtain the best solution.
+
+These ideas require a particular (possibly new) cluster-api cloud provider controller
+which can inspect machinesets and annotate them as appropriate.  This could be
+done dynamically via billing API (not recommended) or have values compiled in
+statically.
+
+We could also attempt to build this logic into the machineset/machinedeployment
+controller, but this would probably not be very scalable and would require
+a well defined set of fields on the provider CRD.
+
+## Motivation
+
+### Goals
+
+### Non-Goals
+
+## Proposal
+
+### User Stories
+
+As a cluster admin, I want to add the ability to automatically scale my cluster
+using cluster autoscaler and the cluster-api.  In order to save money, I need
+the ability to scale from zero for a given machineset or machinedeployment.
+
+## Proposal Ideas
+
+There are several proposals to consider.  Once we decide on one, we'll eliminate
+the others.
+
+### Annotations
+Modify the autoscaler to look at particular annotations on a given object to determine the appropriate amounts of each resource.
+
+#### Benefits
+Easy, simple.  Users can modify if needed (we can instruct controllers to only
+add annotations if they're missing, etc), users can bring their own clouds
+and annotate their objects to support scale from zero and we don't need to worry
+about supporting every cloud day 1.
+
+#### Cons
+We're essentially creating an API around an annotation, and the api folks don't
+like that type of thing.
+
+### Spec Field
+We can add fields to the spec of a particular object to set these items.
```

Comment:
I note the clever ambiguity in "particular object" - were you thinking of the MachineSet/MachineDeployment?  We could also do a "sidecar" object, similar to how we did the `infrastructureRef`.  We could then also put additional autoscaling controls here e.g.

```
kind: MachineDeploymentAutoscaling
spec:
  maxReplicas: 10
  minReplicas: 0
  scaleDownDelay: 10m
  scaleUpDelay: 0s
status:
  nodeShape:
     memory: 1Gi
     gpus: 1
     labels:
      foo: bar
```

I'm not sure where we're currently putting these?  Also it's a little awkward because I would imagine I'd want to set these on the MachineDeployment level, but I'd imagine node-attributes should technically be on the MachineSet level.

Reply:
One out of the box idea I had that seems related to this one was removing the integration between the cluster autoscaler and the cluster-api bits.

We could come up with a new 'external scale group' object.  The autoscaler increments or decrements that object, and something on the cluster-api watches that object and takes the appropriate actions.  This would allow people to build their own automation after the recommendations of the autoscaler, but it's not entirely clear what this should look like.

Reply:
`minReplicas / maxReplicas` are known annotations which apply to any scalable resource i.e machineSet/machineDeployment which makes them discoverable for the cluster autoscaler. They could be owned by. a e.g `machineAutoscaler` resource:
https://github.com/kubernetes-sigs/cluster-api/issues/2369

`cluster.x-k8s.io/cluster-api-autoscaler-node-group-min-size`
`cluster.x-k8s.io/cluster-api-autoscaler-node-group-max-size`

```
apiVersion: "cluster.x-k8s.io"
kind: "MachineAutoscaler"
metadata:
  name: "worker-us-east-1a" 
  namespace: "cluster-api"
spec:
  minReplicas: 1 
  maxReplicas: 12 
  scaleTargetRef: 
    apiVersion: cluster.x-k8s.io
    kind: MachineSet 
    name: worker-us-east-1a
```

>I would imagine I'd want to set these on the MachineDeployment level,

I would expect this info to be accessible through any scalable resource liable to be autoscaled but decoupled from it. This is purely provider specific compute info which does not necessarily needs a replicas controller on top to be meaningful. In this line I'd see the current `providerMachineTemplates` or something like a `machineClass` good candidates to infer and expose this information as they see fit for each specific provider.